Content.Shared/Damage/DamageableSystem.cs

```diff
@@ -0,0 +1,322 @@
+        ///     resulting dictionary.
+        /// </remarks>
+        public Dictionary<string, int> GetDamagePerGroup(Dictionary<string, int> damagePerType)
+        {
+            var damageGroupDict = new Dictionary<string, int>();
+            foreach (var group in _prototypeManager.EnumeratePrototypes<DamageGroupPrototype>())
+            {
+                var groupDamage = 0;
+                var groupIsSupported = false;
+
+                foreach (var type in group.DamageTypes)
+                {
+                    if (damagePerType.TryGetValue(type, out var damage))
+                    {
+                        groupIsSupported = true;
+                        groupDamage += damage;
+                    }
+                }
+
+                // was at least one member of this group actually supported by the container?
+                if (groupIsSupported)
+                {
+                    damageGroupDict.Add(group.ID, groupDamage);
+                }
+            }
+            return damageGroupDict;
+        }
+
+        private void DamageableGetState(EntityUid uid, DamageableComponent component, ref ComponentGetState args)
+        {
+            args.State = new DamageableComponentState(component.DamagePerType, component.ResistanceSetID);
+        }
+
+        private void DamageableHandleState(EntityUid uid, DamageableComponent component, ref ComponentHandleState args)
+        {
+            if (args.Current is not DamageableComponentState state)
+            {
+                return;
+            }
+
+            component.ResistanceSetID = state.ResistanceSetID;
+
+            // Check if the damage per type has updated. Usually a damage change is apparent by just looking at the
+            // total. Only check every single damage type if we need to.
+            var newTotalDamage = state.DamagePerType.Values.Sum();
+            if (component.TotalDamage == newTotalDamage &&
+                component.DamagePerType.Count == state.DamagePerType.Count)
+            {
+                // Check every damage type for changes
+                var damageChanged = false;
+                foreach (var (type, newValue) in state.DamagePerType)
+                {
+                    if (!component.DamagePerType.TryGetValue(type, out var oldValue) ||
+                        oldValue != newValue)
+                    {
+                        damageChanged = true;
+                        break;
+                    }
+                }
+
+                if (!damageChanged)
+                    return;
+            }
+
+            component.DamagePerType = state.DamagePerType;
+
+            // Calculate dependent values and raise a local event. The event is needed as there may be client-exclusive
+            // systems (e.g. UI) that need to know if damage changed as a result of server-exclusive damage-dealing
+            // systems.
+            DamageChanged(component, component.TotalDamage < newTotalDamage);
```

Comment:
So here, this is raising `DamageChangedEvents` inside of the handle damageable component state function.  This feels wrong, but seems like the most reasonable way to ensure that client-side systems are informed when server-side systems deal damage.

Currently this doesn't really matter. Things like MobState are `Dirty()`-ed by the server. But it might matter for some future client-exclusive systems?